appsmithorg/appsmith
# feat: updatedRow property should point to the last edited row in table widget

## Description

This PR updates the logic of how the `updatedRow` derived property is computed. Previously, `updatedRow` was equivalent to `triggeredRow`.  

* ### Intended behaviour:
    * `updatedRow` should be available as soon as a user updates a table cell. Not edits, but updates.
    * `updatedRow` when the mode is set to custom should show data of the most recent row which is updated.
    * The structure of `updatedRow` should be similar to that of `triggeredRow` property. 
    * When there is no change `updatedRow` should be equal to an object with keys having blank string value.

* ### Summary of changes:
    * Introduced a new meta property named: `updatedRowIndex`. This index value helps us to get updated row from the `filteredTableData` or `processTableData`
    * `updatedRowIndex` is updated when the transient table data is updated.
    * It is reset to `-1` whenever:
        * Whenever the value is discarded with discard button. Refer to the `removeRowFromTransientTableData`
        * Whenever the table data is modified. Refer to the `componentDidUpdate`

Fixes #15486

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Below is the list of test cases that is executed to verify my changes:
* **Cypress**:
    * Added below cypress tests to `inline_editing_spec.js` to check the following behaviour for single and multi row update mode:
        * check if updatedRowIndex has -1 as the default value (Only for single row update mode)
        * check if updatedRowIndex is 0, when cell at row 0 is updated (Only for single row update mode)
        * check if updatedRowIndex is -1 when changes are discarded
        * check if the updateRowIndex is -1 when widget is reset
        * check if the updatedRowIndex changes to -1 when the table data changes.
* **Unit test**:
    * Added tests in `derived.test.js` with the following scenarios to validate `getUpdatedRow`:
        * with valid updated row index
        * with updated row index as -1
        * with string as a updated row index
        * with undefined as updated row index
        * for removal of non data columns.
* **Manual**
    * Manually tested that, on click of `onSave` button the `updatedRowIndex` is reset to -1.


### Test Plan
> https://github.com/appsmithorg/TestSmith/issues/2091

### Issues raised during DP testing
> Link issues raised during DP testing for better visiblity and tracking (copy link from comments dropped on this PR)


## Checklist:
### Dev activity


### QA activity:
